File: code/sdk/android/cloudsheet-Android-sdk/src/main/kotlin/com/opencloudsheet/utilities/OneDriveResponseHelper.kt
package com.opencloudsheet.utilities

import com.google.gson.Gson
import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonParser
import com.opencloudsheet.annotations.ColumnIndex
import com.opencloudsheet.model.worksheet.IWorksheetRow

object OneDriveResponseHelper {
    private val _gson: Gson by lazy {
        Gson()
    }

    fun toString(obj: Any): String = _gson.toJson(obj)

    fun parse(response: String): JsonElement? {
        return JsonParser.parseString(response)
    }

    fun <T> fromJson(json: String, clazz: Class<T>): T = _gson.fromJson(json, clazz)

    fun <T : IWorksheetRow> parseTableRows(jsonArray: JsonArray?, clazz: Class<T>): List<T> {
        if (jsonArray == null) {
            return emptyList()
        }
        val result = mutableListOf<T>()
        for (i in 0 until jsonArray.size()) {
            val rowObject = jsonArray[i].asJsonObject
            val parsedRow = parseRow(rowObject, clazz, i)
            result.add(parsedRow)
        }
        return result
    }

    fun <T : IWorksheetRow> parseRow(rowObject: com.google.gson.JsonObject, clazz: Class<T>, fallbackIndex: Int = 0): T {
        val valuesArray = rowObject.getAsJsonArray("values")
        val rowValues = valuesArray[0].asJsonArray

        // Get the actual row index from rowObject
        val rowIndex = rowObject.get("index")?.asInt ?: fallbackIndex

        val jsonObject = buildJsonObjectForDeserialization(rowValues, clazz)
        val instance = _gson.fromJson(jsonObject, clazz)
        instance.setRowIndex(rowIndex)
        return instance
    }

    /**
     * Build a JsonObject from valuesArray for Gson deserialization.
     * Uses OneDriveWorksheetRowHelpers.buildColumnMapping to get field mapping.
     */
    private fun <T : IWorksheetRow> buildJsonObjectForDeserialization(
        valuesArray: JsonArray,
        clazz: Class<T>
    ): com.google.gson.JsonObject {
        // Get complete column mapping (index -> field name)
        val columnMapping = OneDriveWorksheetRowHelpers.buildColumnMapping(clazz)

        // Build JSON object from valuesArray using column mapping
        val jsonObject = com.google.gson.JsonObject()
        columnMapping.forEach { (columnIndex, fieldName) ->
            if (columnIndex < valuesArray.size()) {
                val value = valuesArray[columnIndex]
                jsonObject.add(fieldName, value)
            }
        }

        return jsonObject
    }

    fun <T : IWorksheetRow> toValueArray(obj: T): List<Any?> {
        val clazz = obj::class.java
        val metadataColumnCount = 3
        val userFieldNames = obj.getTableColumnFieldsOrder()

        val columnMap = mutableMapOf<Int, Any?>(
            0 to obj.getUniqueRowId(),
            1 to obj.getCreatedAt(),
            2 to obj.getUpdatedAt()
        )

        userFieldNames.forEachIndexed { index, fieldName ->
            val field = clazz.getDeclaredField(fieldName)
            field.isAccessible = true
            val actualColumnIndex = index + metadataColumnCount
            columnMap[actualColumnIndex] = field.get(obj)
        }

        val maxIndex = columnMap.keys.maxOrNull() ?: -1
        return (0..maxIndex).map { columnMap[it] }
    }
}

File: code/sdk/ios/OpenCloudSheet/Sources/OpenCloudSheet/Utilities/OneDriveResponseHelper.swift
//
//  OneDriveResponseHelper.swift
//  OpenCloudSheet
//
//  Helper for parsing OneDrive API responses
//

import Foundation

class OneDriveResponseHelper {
    static func parse(_ response: String) -> [String: Any]? {
        guard let data = response.data(using: .utf8) else { return nil }
        return try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    }

    static func parseTableRows<T: IWorksheetRow>(
        _ jsonArray: [[String: Any]]?,
        _ type: T.Type
    ) -> [T] {
        guard let jsonArray = jsonArray else { return [] }

        var result: [T] = []
        for (index, rowObject) in jsonArray.enumerated() {
            if let parsedRow = parseRow(rowObject, type, fallbackIndex: index) {
                result.append(parsedRow)
            }
        }
        return result
    }

    static func parseRow<T: IWorksheetRow>(
        _ rowObject: [String: Any],
        _ type: T.Type,
        fallbackIndex: Int = 0
    ) -> T? {
        guard let valuesArray = rowObject["values"] as? [[Any]] else { return nil }
        guard let rowValues = valuesArray.first else { return nil }

        let rowIndex = rowObject["index"] as? Int ?? fallbackIndex

        var instance = buildInstanceFromValues(rowValues, type)
        instance._rowIndex = rowIndex
        return instance
    }

    private static func buildInstanceFromValues<T: IWorksheetRow>(
        _ valuesArray: [Any],
        _ type: T.Type
    ) -> T {
        let columnMapping = OneDriveWorksheetRowHelpers.buildColumnMapping(type)
                        
        var jsonDict: [String: Any] = [:]
        jsonDict["_rowIndex"] = -1
        for (columnIndex, fieldName) in columnMapping {
            guard columnIndex < valuesArray.count else { continue }
            let value = valuesArray[columnIndex]
            jsonDict[fieldName] = value
        }

        do {
            let jsonData = try JSONSerialization.data(withJSONObject: jsonDict)
            let instance = try JSONDecoder().decode(type, from: jsonData)
            return instance
        } catch {
            print("Failed to deserialize row: \(error)")
            fatalError("Failed to deserialize row of type \(type)")
        }
    }


    static func toValueArray<T: IWorksheetRow>(_ obj: T) -> [Any?] {
        let metadataColumnCount = 3
        let userFieldNames = type(of: obj).getTableColumnFieldsOrder()

        var columnMap: [Int: Any?] = [
            0: obj._rowId,
            1: obj._createdAt,
            2: obj._updatedAt
        ]

        let mirror = Mirror(reflecting: obj)
        for (index, fieldName) in userFieldNames.enumerated() {
            let actualColumnIndex = index + metadataColumnCount
            if let child = mirror.children.first(where: { $0.label == fieldName }) {
                columnMap[actualColumnIndex] = child.value
            }
        }

        let maxIndex = columnMap.keys.max() ?? -1
        return (0...maxIndex).map { columnMap[$0] ?? NSNull() }
    }
}
